README.md
[![Build status](https://ci.appveyor.com/api/projects/status/0y4lccfjpm8s5bgg?svg=true)](https://ci.appveyor.com/project/Chris00/root1d)

Root1D — Find roots of 1D functions
===================================

The module `Root1D` provides a collection of functions to seek roots
of functions `float → float`.


Installation
------------

The easier way of installing this package is by using [opam][]:

```shell
opam install root1d
```

To compile by hand, install [dune][] and do `dune build @install`.


[opam]: https://opam.ocaml.org/
[dune]: https://github.com/ocaml/dune

Documentation
-------------

See the [signature of `Root1D`](src/Root1D.mli).  It can also be
consulted rendered to [HTML](https://chris00.github.io/root1d/doc/).


CHANGES.md
0.5.1 2018-09-07
----------------

- Use [Dune](https://github.com/ocaml/dune) and
  [dune-release](https://github.com/samoht/dune-release).
- Publish the documentation on Github.

0.5 2017-08-14
--------------

- Use jbuilder and Topkg.
- Changed license to ISC.


tests/speed.ml
open Format
open Benchmark

(* Internal function for brent's method *)
let rec brent4_int tol f a fa b fb c fc mflag d =
  let s =
    if fb <> fc && fa <> fc then (
      (* inverse quadratic interpolation *)
      a *. fb *. fc /. (fa -. fb) /. (fa -. fc)
      +. b *. fa *. fc /. (fb -. fa) /. (fb -. fc)
      +. c *. fa *. fb /. (fc -. fa) /. (fc -. fb)
    ) else ((* secant rule *)
      b -. fb *. (b -. a) /. (fb -. fa)
    )
  in
  (* condition 1-5 to reject above and use bisection instead *)
  let delta = epsilon_float *. abs_float b +. tol in (* ??? *)
  let s, mflag =
    if (if a < b
      then s < (3. *. a +. b) *. 0.25 || s > b
      else s > (3. *. a +. b) *. 0.25 || s < b)
      || (if mflag then
          abs_float (s -. b) >= abs_float (b -. c) *. 0.5
          || abs_float (b -. c) < delta
        else
          abs_float (s -. b) >= abs_float (c -. d) *. 0.5
          || abs_float (c -. d) < delta
      )
    then (a +. b) *. 0.5, true else s, false
  in
  let fs = f s in
  if fs = 0. || abs_float (b -. a) < epsilon_float *. abs_float b +. tol then s
  else
    if fa *. fs < 0. then
      brent4_int_swap tol f a fa s fs b fb mflag c
    else
      brent4_int_swap tol f s fs b fb b fb mflag c

(* helper for a-b swapping and xdelts checks *)
and brent4_int_swap tol f a fa b fb c fc mflag d =
  (* finish rootfinding if our range is smaller than xdelta *)
  if abs_float (b -. a) < epsilon_float *. abs_float b +. tol then b
  else
    (* ensure that fb is the best estimate so far by swapping b with a *)
    if abs_float fa < abs_float fb then
      brent4_int tol f b fb a fa c fc mflag d
    else
      brent4_int tol f a fa b fb c fc mflag d

let eps = sqrt epsilon_float

(* optimized version 2 of brent for performance comparison *)
let brent4 ?(tol=eps) f a b =
  let fa = f a in
  let fb = f b in
  if fa *. fb >= 0. then failwith "Root must be bracketed";
  brent4_int_swap tol f a fa b fb a fa true 0.


let f x = 4. *. x *. x *. x -. 16. *. x *. x +. 17. *. x -. 4.
(* Actual roots:
0.3285384586114149
1.2646582900644197
2.4068032513241651
 *)

let f1 x = (x +. 3.) *. (x -. 1.) *. (x -. 1.)
(* roots: -3, 1 (double root) *)

let f2 x = tan x -. 2. *.  x
(* root: 1.16556118520721 *)

let f3 x = x**19.

let f4 x = if x = 0. then 0. else x *. exp(-1. /. x)
let f4' x = if x = 0. then 0. else x *. exp(-1. /. sqrt(abs_float x))

(* brent4 fails for delta = 1e-3 *)
let delta = 2e-3 and a = 0. and b = 1.
let f5_a = -. (b -. a -. delta) /. delta *. 2.**(b /. delta)
let f5 x =
  if a +. delta <= x && x <= b then 2.**(x /. delta)
  else if x = a then f5_a
  else f5_a +. (2.**(a /. delta) -. f5_a) *. (x -. a) /. delta (* arbitrary *)

let print get_root =
  let get_root f a b =
    let n = ref 0 in
    let x = get_root (fun x -> incr n; f x) a b in
    x, !n in
  List.iter (fun (f_name, f,a,b) ->
             let x, n = get_root f a b in
             printf "%s: %f [%i] @ " f_name x n)
            ["f", f, 0., 1.; "f", f, 1., 2.; "f", f, 2., 3.;
             "f1", f1, (-4.), (4./.3.);  "f2", f2, 0.5, 1.5;
             "f3", f3, -1., 4.;  "f4", f4, -1., 4.;  "f5", f5, 0., 1.]

let () =
  printf "Bisection:@\n";
  print Root1D.bisection;
  printf "Brent:@\n";
  print Root1D.brent;
  printf "Brent 4:@\n";
  print brent4;
  printf "@\n%!"

let bench get_root () =
  let _ = get_root f 0. 1. in
  let _ = get_root f 1. 2. in
  let _ = get_root f 2. 3. in
  let _ = get_root f1 (-4.) (4./.3.) in
  let _ = get_root f2 0.5 1.5 in
  (* let _ = get_root f3 (-1.) 4. in *)
  let _ = get_root f4 (-1.) 4. in
  let _ = get_root f5 0. 1. in
  ()

let t = throughputN 1 ~repeat:5
                    ["Root1D", bench Root1D.brent, ();
                     "brent4", bench brent4, ()]

let () =
  tabulate t


(* Local Variables: *)
(* compile-command: "make -k -C .." *)
(* End: *)


tests/difficult.ml
(* Some functions for which roots are difficult to compute.  Inspired
   from examples in Brent, R. (1973) Algorithms for Minimization
   without Derivatives. Englewood Cliffs, NJ: Prentice-Hall. *)

open Printf

let () =
  let delta = 1e-4
  and a = 0. and b = 1. in
  let fa = -. (b -. a -. delta) /. delta *. 2.**(b /. delta) in
  let f_a_delta = 2.**(a /. delta) in
  let root = a -. delta *. (f_a_delta /. fa -. 1.) in
  let f x =
    if a +. delta <= x && x <= b then 2.**(x /. delta)
    else if x = a then fa
    else fa +. (f_a_delta -. fa) *. (x -. a) /. delta (* cont. extension *)
  in
  let x = Root1D.brent f a b in
  printf "Root: %f (err: %g)\n" x (x -. root)

let () =
  (* All derivatives vanish at the root x=0. *)
  let f x = if x = 0. then 0. else x *. exp(-1. /. (abs_float x)**0.1) in
  printf "Root: %g (exact: 0)\n" (Root1D.brent f (-1.) 4.)


tests/precision.ml
(* Compare the various methods of root finding. *)

open Printf

let pi = acos(-1.)

let f a x = cos x -. a *. sqrt (sin x)

let df a x = -. sin x -. 0.5 *. a *. cos x /. sqrt (sin x)

let compare methods a =
  let get_root meth =
    let n = ref 0 in (* number of fun eval *)
    let x = meth (fun x -> incr n; f a x) 0. pi in
    !n, x in
  match List.map get_root methods with
  | [] -> ()
  | (n0, x0) :: sols ->
     (* Use [x0] as the reference solution. *)
     printf "a=%2g: %g [%d]\t" a x0 n0;
     List.iter (fun (n, x) ->
                let e = abs_float((x -. x0) /. x0) in
                printf "%g [%d] ε=%.2g\t" x n e
               ) sols;
     printf "\n"

let () =
  let m = [(fun f a b -> Root1D.bisection f a b ~eps:1e-14);
           (fun f a b -> Root1D.brent f a b ~tol:1e-30);
           (fun f a b -> Root1D.illinois f a b ~eps:1e-14);
          ] in
  printf "      Bisection [#eval]\tBrent\n";
  List.iter (compare m) [1.; 2.; 3.; 4.; 5.; 10.; 20.; 40.; 60.; 100.;
                         300.; 500.; 1000.; 5000.; 10000.; 50000.]


src/root1D.ml
(* File: RootFinding.ml

   Copyright (C) 2007

     Christophe Troestler
     email: Christophe.Troestler@umh.ac.be
     WWW: http://math.umh.ac.be/an/software/

   This library is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License version 2.1 or
   later as published by the Free Software Foundation, with the special
   exception on linking described in the file LICENSE.

   This library is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the file
   LICENSE for more details. *)

open Printf

let eps = sqrt epsilon_float
let half_eps = 0.5 *. eps

exception Root of float
(* Internal exception used when the exact root is found *)


(* Assume fa = f(a) < 0 < fb = f(b).  A recursive function is more
   elegant but we do it imperatively to allow the compiler to unbox
   the floats. *)
let do_bisection good_enough f a b fa fb =
  let a = ref a and b = ref b
  and fa = ref fa and fb = ref fb in
  try
    while not(good_enough !a !b !fa !fb) do
      let m = !a +. 0.5 *. (!b -. !a) in
      let fm = f m in
      if fm = 0. then raise(Root m)
      else if fm < 0. then (a := m; fa := fm)
      else (* fm > 0. *) (b := m; fb := fm)
    done;
    !a +. 0.5 *. (!b -. !a)
  with Root r -> r

let bisection_good a b _ _ =
  abs_float(a -. b) < half_eps *. (abs_float a +. abs_float b)

let bisection ?(good_enough=bisection_good) f a b =
  let fa = f a
  and fb = f b in
  if fa = 0. then a
  else if fa < 0. then
    if fb = 0. then b
    else if fb < 0. then
      invalid_arg "Root1D.bisection: f(a) and f(b) are both < 0."
    else (* fb > 0. *)
      do_bisection good_enough f a b fa fb
  else (* fa > 0. *)
    if fb = 0. then b
    else if fb > 0. then
      invalid_arg "Root1D.bisection: f(a) and f(b) are both > 0."
    else (* fb < 0. *)
      do_bisection good_enough f b a fb fa

let newton_good x xpre fx = abs_float fx < eps

let newton ?(good_enough=newton_good) f_f' x0 =
  let x = ref x0
  and xpre = ref nan (* FIXME: confusing for good_enough *)
  and fx, f'x = f_f' x0 in
  let fx = ref fx
  and f'x = ref f'x in
  while not(good_enough !x !xpre !fx) do
    if !f'x = 0. then failwith(sprintf "Root1D.newton: f'(%g) = 0" !x);
    x := !x -. !fx /. !f'x;
    let fx_next, f'x_next = f_f' !x in
    fx := fx_next;
    f'x := f'x_next;
  done;
  !x +. 0.


let secant ?(good_enough) f a b =
  a

let false_position ?(good_enough) f a b =
  a

let muller f a b = a



(* Based on http://www.netlib.org/c/brent.shar zeroin  and
   http://www.netlib.org/go/zeroin.f

   Algorithm

   G.Forsythe, M.Malcolm, C.Moler, Computer methods for mathematical
   computations. M., Mir, 1980, p.180 of the Russian edition

   The function makes use of the bissection procedure combined with
   the linear or quadric inverse interpolation.
   At every step program operates on three abscissae - a, b, and c.
   b - the last and the best approximation to the root
   a - the last but one approximation
   c - the last but one or even earlier approximation than a that
       1) |f(b)| <= |f(c)|
       2) f(b) and f(c) have opposite signs, i.e. b and c confine
     	  the root
   At every step Zeroin selects one of the two new approximations, the
   former being obtained by the bissection procedure and the latter
   resulting in the interpolation (if a,b, and c are all different
   the quadric interpolation is utilized, otherwise the linear one).
   If the latter (i.e. obtained by the interpolation) point is
   reasonable (i.e. lies within the current interval [b,c] not being
   too close to the boundaries) it is accepted. The bissection result
   is used in the other case. Therefore, the range of uncertainty is
   ensured to be reduced at least by the factor 1.6

   See also www.physics.mcgill.ca/~patscott/teaching/numeric/Lec%203.pdf *)
let brent ?(tol=eps) f a0 b0 =
  let a = ref a0
  and b = ref b0
  and c = ref a0 in
  let fa = ref(f !a)
  and fb = ref(f !b) in
  let fc = ref(!fa) in
  if (!fa < 0. && !fb < 0.) || (!fa > 0. && !fb > 0.) then
    invalid_arg "Root1D.brent: f(a) and f(b) must have opposite signs";
  try
    while true do
      let prev_step = !b -. !a in
      (* Swap b and c for b to be the best approximation *)
      if abs_float !fc < abs_float !fb then (
        a := !b;   b := !c;   c := !a;
        fa := !fb; fb := !fc; fc := !fa;
      );
      let tol_act = 2. *. epsilon_float *. abs_float(!b) +. 0.5 *. tol in
      let bisec_step = 0.5 *. (!c -. !b) in
      if abs_float bisec_step <= tol_act || !fb = 0. then
        raise(Root !b);
      let new_step =
        if abs_float prev_step >= tol_act
           && abs_float !fa > abs_float !fb then
          (* prev_step was large enough and was in true direction,
             Interpolatiom may be tried *)
          let c_b = !c -. !b in
          let p, q =
            if !a = !c then
              (* linear interpolation *)
              let s = !fb /. !fa in (c_b *. s, 1. -. s)
            else
              (* Quadric inverse interpolation *)
              let t = !fa /. !fc and r = !fb /. !fc and s = !fb /. !fa in
              (s *. (c_b *. t *. (t -. r) -. (!b -. !a) *. (r -. 1.)),
               (t -. 1.) *. (r -. 1.) *. (s -. 1.)) in
          let p, q = if p > 0. then p, -. q else -. p, q in
          (* If b+p/q falls in [b,c] and isn't too large, it is accepted *)
          if p < 0.75 *. c_b *. q -. 0.5 *. abs_float(tol_act *. q)
             && p < abs_float(0.5 *. prev_step *. q) then p /. q
          else bisec_step
        else bisec_step in
      (* Adjust the step to be not less than tolerance *)
      let new_step =
        if new_step > 0. then
          if new_step < tol_act then tol_act else new_step
        else
          if -. new_step < tol_act then -. tol_act else new_step in
      a := !b;  fa := !fb; (* Save the previous approx. *)
      b := !b +. new_step;
      fb := f(!b);
      (* Adjust c for it to have a sign opposite to that of b *)
      if (!fb > 0. && !fc > 0.) || (!fb < 0. && !fc < 0.) then (
        c := !a;  fc := !fa
      )
    done;
    assert false
  with Root r -> r


let twice_epsilon_float = 2. *. epsilon_float

let rec brent_loop half_tol f a fa b fb c fc d e =
  (* [b]: best guess for the root, |f(b)| ≤ |f(a)|, |f(c)|.
     [c]: opposite side of x axis to [b], so [b] and [c] bracket the root.
     [a]: previous best guess.
   *)
  let tol_act = twice_epsilon_float *. abs_float(b) +. half_tol in
  let m = 0.5 *. (c -. b) in
  if abs_float m <= tol_act || fb = 0. then b
  else (
    let step, e' =
      if abs_float e < tol_act || abs_float fa <= abs_float fb then
        m, m (* bisection *)
      else
        (* prev_step was large enough and was in true direction,
           Interpolatiom may be tried *)
        let s = fb /. fa in
        let p, q =
          if a = c then (* Linear interpolation *)
            (2. *. m *. s, 1. -. s)
          else
            (* Inverse quadratic interpolation *)
            let q = fa /. fc and r = fb /. fc in
            (s *. (2. *. m *. q *. (q -. r) -. (b -. a) *. (r -. 1.)),
             (q -. 1.) *. (r -. 1.) *. (s -. 1.)) in
        let p, q = if p > 0. then p, -. q else -. p, q in
        (* If b+p/q falls in [b,c] and isn't too large, it is accepted *)
        if 2. *. p < 3. *. m *. q -. abs_float(tol_act *. q)
           && p < abs_float(0.5 *. e *. q)
        then p /. q, d
        else m, m (* bisection *)
    in
    (* Adjust the step to be not less than tolerance *)
    let b' = b +. (if abs_float step > tol_act then step
                   else if m > 0. then tol_act else -. tol_act) in
    let fb' = f b' in
    (* Adjust c for it to have a sign opposite to that of b *)
    if (fb' > 0.) = (fc > 0.) then (* => fb' * fb <= 0 *)
      let d = b' -. b in
      if abs_float fb < abs_float fb' then
        brent_loop half_tol f b' fb' b fb b' fb' d d
      else
        brent_loop half_tol f b fb b' fb' b fb d d
    else (* => fb' * fc <= 0 *)
      if abs_float fc < abs_float fb' then
        brent_loop half_tol f b' fb' c fc b' fb' step e'
      else
        brent_loop half_tol f b fb b' fb' c fc step e'
  )
;;

let brent ?(tol=eps) f a b =
  if tol < 0. then invalid_arg "Root1D.brent: tol < 0.";
  let fa = f a and fb = f b in
  if fa = 0. then a
  else if fb = 0. then b
  else if (fa < 0. && fb < 0.) || (fa > 0. && fb > 0.) then
    invalid_arg "Root1D.brent: f(a) and f(b) must have opposite signs"
  else
    let d = b -. a in
    if abs_float fa < abs_float fb then
      brent_loop (0.5 *. tol) f b fb a fa b fb d d
    else
      brent_loop (0.5 *. tol) f a fa b fb a fa d d


let rec brent2_loop half_tol f  a fa ea  b fb eb  c fc ec  d e =
  let tol_act = twice_epsilon_float *. abs_float(b) +. half_tol in
  let m = 0.5 *. (c -. b) in
  if abs_float m <= tol_act || fb = 0. then b
  else (
    let step, e' =
      if abs_float e < tol_act
         || (ea <= eb && ldexp (abs_float fa) (ea - eb) <= abs_float fb)
         || (ea > eb && ldexp (abs_float fb) (eb - ea) >= abs_float fa) then
        m, m
      else
        let s = ldexp fb (eb - ea) /. fa in
        let p, q =
          if a = c then (* Linear interpolation *)
            (2. *. m *. s, 1. -. s)
          else
            (* Inverse quadratic interpolation *)
            let q = ldexp fa (ea - ec) /. fc
            and r = ldexp fb (eb - ec) /. fc in
            (s *. (2. *. m *. q *. (q -. r) -. (b -. a) *. (r -. 1.)),
             (q -. 1.) *. (r -. 1.) *. (s -. 1.)) in
        let p, q = if p > 0. then p, -. q else -. p, q in
        (* If b+p/q falls in [b,c] and isn't too large, it is accepted *)
        if 2. *. p < 3. *. m *. q -. abs_float(tol_act *. q)
           && p < abs_float(0.5 *. e *. q)
        then p /. q, d
        else m, m
    in
    (* Adjust the step to be not less than tolerance *)
    let b' = b +. (if abs_float step > tol_act then step
                   else if m > 0. then tol_act else -. tol_act) in
    let fb', eb' = f b' in
    (* Adjust c for it to have a sign opposite to that of b *)
    if (fb' > 0.) = (fc > 0.) then (* => fb' * fb <= 0 *)
      let d = b' -. b in
      if (eb <= eb' && ldexp (abs_float fb) (eb - eb') < abs_float fb')
         || (eb > eb' && ldexp (abs_float fb') (eb' - eb) >= abs_float fb) then
        brent2_loop half_tol f b' fb' eb' b fb eb b' fb' eb' d d
      else
        brent2_loop half_tol f b fb eb b' fb' eb' b fb eb d d
    else (* => fb' * fc <= 0 *)
      if (ec <= eb' && ldexp (abs_float fc) (ec - eb') < abs_float fb')
         || (ec > eb' && ldexp (abs_float fb') (eb' - ec) >= abs_float fc) then
        brent2_loop half_tol f b' fb' eb' c fc ec b' fb' eb' step e'
      else
        brent2_loop half_tol f b fb eb b' fb' eb' c fc ec step e'
  )

let brent2 ?(tol=eps) f a b =
  if tol < 0. then invalid_arg "Root1D.brent2: tol < 0.";
  let fa, ea = f a and fb, eb = f b in
  if fa = 0. then a
  else if fb = 0. then b
  else if (fa < 0. && fb < 0.) || (fa > 0. && fb > 0.) then
    invalid_arg "Root1D.brent: f(a) and f(b) must have opposite signs"
  else
    let d = b -. a in
    if (ea <= eb && ldexp (abs_float fa) (ea - eb) < abs_float fb)
       || (ea > eb && ldexp (abs_float fb) (eb - ea) >= abs_float fa) then
      brent2_loop (0.5 *. tol) f  b fb eb  a fa ea  b fb eb  d d
    else
      brent2_loop (0.5 *. tol) f  a fa ea  b fb eb  a fa ea  d d


(* Local Variables: *)
(* compile-command: "make -k -C .." *)
(* End: *)


src/Root1D.ml
(* File: Root1D.ml

   Copyright (C) 2007

     Christophe Troestler
     email: Christophe.Troestler@umons.ac.be
     WWW: http://math.umons.ac.be/anum/software/

   This library is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License version 2.1 or
   later as published by the Free Software Foundation, with the special
   exception on linking described in the file LICENSE.

   This library is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the file
   LICENSE for more details. *)

open Printf

let eps = sqrt epsilon_float

exception Root of float
(* Internal exception used when the exact root is found *)

let max_float (a: float) (b: float) =
  if a >= b then a (* ≠ NaN *)
  else b

(* Assume fa = f(a) < 0 < fb = f(b).  A recursive function is more
   elegant but we do it imperatively to allow the compiler to unbox
   the floats. *)
let do_bisection improve f a b fa fb =
  let a = ref a and b = ref b
  and fa = ref fa and fb = ref fb in
  try
    while improve !a !b do
      let m = !a +. 0.5 *. (!b -. !a) in
      let fm = f m in
      if fm = 0. then raise(Root m)
      else if fm < 0. then (a := m; fa := fm)
      else (* fm > 0. *) (b := m; fb := fm)
    done;
    !a +. 0.5 *. (!b -. !a)
  with Root r -> r

let bisection_improve_default a b =
  abs_float(a -. b) > eps *. max_float (abs_float a) (abs_float b)

let bisection_improve_eps eps a b =
  abs_float(a -. b) > eps *. max_float (abs_float a) (abs_float b)

let bisection ?eps f a b =
  let improve = match eps with
    | None -> bisection_improve_default
    | Some eps ->
       if eps <= 0. then invalid_arg "Root1D.bisection: tol <= 0";
       bisection_improve_eps eps in
  let fa = f a
  and fb = f b in
  if fa = 0. then a
  else if fa < 0. then
    if fb = 0. then b
    else if fb < 0. then
      invalid_arg "Root1D.bisection: f(a) and f(b) are both < 0."
    else (* fb > 0. *)
      do_bisection improve f a b fa fb
  else (* fa > 0. *)
    if fb = 0. then b
    else if fb > 0. then
      invalid_arg "Root1D.bisection: f(a) and f(b) are both > 0."
    else (* fb < 0. *)
      do_bisection improve f b a fb fa

let newton_good x xpre fx = abs_float fx < eps

let newton ?(good_enough=newton_good) f_f' x0 =
  let x = ref x0
  and xpre = ref nan (* FIXME: confusing for good_enough *)
  and fx, f'x = f_f' x0 in
  let fx = ref fx
  and f'x = ref f'x in
  while not(good_enough !x !xpre !fx) do
    if !f'x = 0. then failwith(sprintf "Root1D.newton: f'(%g) = 0" !x);
    x := !x -. !fx /. !f'x;
    let fx_next, f'x_next = f_f' !x in
    fx := fx_next;
    f'x := f'x_next;
  done;
  !x *. 1.


type last_iter = A | B

(* Assume fa = f(a) < 0 < f(b) = fb. *)
let do_illinois improve f a b fa fb =
  let a = ref a
  and b = ref b in
  let fa = ref fa in
  let fb = ref fb in
  try
    (* One could also implement this using x0 and x1, the last 2
       iterations, but this is slightly slower. *)
    let last = ref B in
    while improve !a !b do
      let x = !b -. !fb *. (!b -. !a) /. (!fb -. !fa) in
      let fx = f x in
      if fx = 0. then raise(Root x);
      match !last with
      | A -> if fx > 0. then (b := x;  fb := fx;  last := B)
             else (a := x;  fa := fx;  fb := 0.5 *. !fb)
      | B -> if fx < 0. then (a := x;  fa := fx;  last := A)
             else (b := x;  fb := fx;  fa := 0.5 *. !fa)
    done;
    match !last with A -> !a | B -> !b
  with Root r -> r

let illinois ?eps f a b =
  let improve = match eps with
    | None -> bisection_improve_default
    | Some eps ->
       if eps <= 0. then invalid_arg "Root1D.bisection: tol <= 0";
       bisection_improve_eps eps in
  let fa = f a in
  if fa = 0. then a
  else if fa < 0. then
    let fb = f b in
    if fb = 0. then b
    else if fb < 0. then
      invalid_arg "Root.illinois: f(a) and f(b) are both < 0."
    else (* fb > 0. *) do_illinois improve f a b fa fb
  else (* fa > 0. *)
    let fb = f b in
    if fb = 0. then b
    else if fb > 0. then
      invalid_arg "Root.illinois: f(a) and f(b) are both > 0."
    else (* fb < 0. *) do_illinois improve f b a fb fa


let muller f a b = a



(* Based on http://www.netlib.org/c/brent.shar zeroin  and
   http://www.netlib.org/go/zeroin.f

   Algorithm

   G.Forsythe, M.Malcolm, C.Moler, Computer methods for mathematical
   computations. M., Mir, 1980, p.180 of the Russian edition

   The function makes use of the bissection procedure combined with
   the linear or quadric inverse interpolation.
   At every step program operates on three abscissae - a, b, and c.
   b - the last and the best approximation to the root
   a - the last but one best approximation
   c - the last but one or even earlier approximation than a that
       1) |f(b)| <= |f(c)|
       2) f(b) and f(c) have opposite signs, i.e. b and c confine
     	  the root
   At every step Zeroin selects one of the two new approximations, the
   former being obtained by the bissection procedure and the latter
   resulting in the interpolation (if a,b, and c are all different
   the quadric interpolation is utilized, otherwise the linear one).
   If the latter (i.e. obtained by the interpolation) point is
   reasonable (i.e. lies within the current interval [b,c] not being
   too close to the boundaries) it is accepted. The bissection result
   is used in the other case. Therefore, the range of uncertainty is
   ensured to be reduced at least by the factor 1.6

   See also www.physics.mcgill.ca/~patscott/teaching/numeric/Lec%203.pdf *)
let brent ?(tol=eps) f a0 b0 =
  let a = ref a0
  and b = ref b0
  and c = ref a0 in
  let fa = ref(f !a)
  and fb = ref(f !b) in
  let fc = ref(!fa) in
  if !fa = 0. then !a
  else if !fb = 0. then !b
  else if !fa *. !fb > 0. then
    invalid_arg "Root1D.brent: f(a) and f(b) must have opposite signs"
  else (
    let continue = ref true in
    while !continue do
      let prev_step = !b -. !a in
      (* Swap b and c for b to be the best approximation *)
      if abs_float !fc < abs_float !fb then (
        a := !b;   b := !c;   c := !a;
        fa := !fb; fb := !fc; fc := !fa;
      );
      let tol_act = 2. *. epsilon_float *. abs_float(!b) +. 0.5 *. tol in
      let c_b = !c -. !b in
      if 0.5 *. abs_float c_b <= tol_act || !fb = 0. then
        continue := false (* the root is in [b] *)
      else (
        let new_step =
          if abs_float prev_step >= tol_act
             && abs_float !fa > abs_float !fb then
            (* prev_step was large enough and was in true direction,
             Interpolatiom may be tried *)
            let p, q =
              if !a = !c then
                (* linear interpolation *)
                let s = !fb /. !fa in (c_b *. s, 1. -. s)
              else
                (* Quadric inverse interpolation *)
                let t = !fa /. !fc and r = !fb /. !fc and s = !fb /. !fa in
                (s *. (c_b *. t *. (t -. r) -. (!b -. !a) *. (r -. 1.)),
                 (t -. 1.) *. (r -. 1.) *. (s -. 1.)) in
            let p, q = if p > 0. then p, -. q else -. p, q in
            (* If b+p/q falls in [b,c] and isn't too large, it is accepted *)
            if p < 0.75 *. c_b *. q -. 0.5 *. abs_float(tol_act *. q)
               && p < abs_float(0.5 *. prev_step *. q) then p /. q
            else 0.5 *. c_b
          else 0.5 *. c_b in
        a := !b;  fa := !fb; (* Save the previous approx. *)
        if abs_float new_step > tol_act then
          b := !b +. new_step
        else
          (* Adjust the step to be not less than tolerance *)
          b := !b +. copysign tol_act c_b;
        fb := f(!b);
        (* Adjust c for it to have a sign opposite to that of b *)
        if !fb *. !fc > 0. then (
          c := !a;  fc := !fa;
          assert(!fb *. !fc <= 0.);
        )
      )
    done;
    !b
  )

let twice_epsilon_float = 2. *. epsilon_float

let rec brent_loop half_tol f a fa b fb c fc d e =
  (* [b]: best guess for the root, |f(b)| ≤ |f(a)|, |f(c)|.
     [c]: opposite side of x axis to [b], so [b] and [c] bracket the root.
     [a]: previous best guess.
   *)
  let tol_act = twice_epsilon_float *. abs_float(b) +. half_tol in
  let m = 0.5 *. (c -. b) in
  if abs_float m <= tol_act || fb = 0. then b
  else (
    let step, e' =
      if abs_float e < tol_act || abs_float fa <= abs_float fb then
        m, m (* bisection *)
      else
        (* prev_step was large enough and was in true direction,
           Interpolatiom may be tried *)
        let s = fb /. fa in
        let p, q =
          if a = c then (* Linear interpolation *)
            (2. *. m *. s, 1. -. s)
          else
            (* Inverse quadratic interpolation *)
            let q = fa /. fc and r = fb /. fc in
            (s *. (2. *. m *. q *. (q -. r) -. (b -. a) *. (r -. 1.)),
             (q -. 1.) *. (r -. 1.) *. (s -. 1.)) in
        let p, q = if p > 0. then p, -. q else -. p, q in
        (* If b+p/q falls in [b,c] and isn't too large, it is accepted *)
        if 2. *. p < 3. *. m *. q -. abs_float(tol_act *. q)
           && p < abs_float(0.5 *. e *. q)
        then p /. q, d
        else m, m (* bisection *)
    in
    (* Adjust the step to be not less than tolerance *)
    let b' = b +. (if abs_float step > tol_act then step
                   else if m > 0. then tol_act else -. tol_act) in
    let fb' = f b' in
    (* Adjust c for it to have a sign opposite to that of b *)
    if (fb' > 0.) = (fc > 0.) then (* => fb' * fb <= 0 *)
      let d = b' -. b in
      if abs_float fb < abs_float fb' then
        brent_loop half_tol f b' fb' b fb b' fb' d d
      else
        brent_loop half_tol f b fb b' fb' b fb d d
    else (* => fb' * fc <= 0 *)
      if abs_float fc < abs_float fb' then
        brent_loop half_tol f b' fb' c fc b' fb' step e'
      else
        brent_loop half_tol f b fb b' fb' c fc step e'
  )
;;

let brent1 ?(tol=eps) f a b =
  if tol < 0. then invalid_arg "Root1D.brent: tol < 0.";
  let fa = f a and fb = f b in
  if fa = 0. then a
  else if fb = 0. then b
  else if (fa < 0. && fb < 0.) || (fa > 0. && fb > 0.) then
    invalid_arg "Root1D.brent: f(a) and f(b) must have opposite signs"
  else
    let d = b -. a in
    if abs_float fa < abs_float fb then
      brent_loop (0.5 *. tol) f b fb a fa b fb d d
    else
      brent_loop (0.5 *. tol) f a fa b fb a fa d d


let rec brent2_loop half_tol f  a fa ea  b fb eb  c fc ec  d e =
  let tol_act = twice_epsilon_float *. abs_float(b) +. half_tol in
  let m = 0.5 *. (c -. b) in
  if abs_float m <= tol_act || fb = 0. then b
  else (
    let step, e' =
      if abs_float e < tol_act
         || (ea <= eb && ldexp (abs_float fa) (ea - eb) <= abs_float fb)
         || (ea > eb && ldexp (abs_float fb) (eb - ea) >= abs_float fa) then
        m, m
      else
        let s = ldexp fb (eb - ea) /. fa in
        let p, q =
          if a = c then (* Linear interpolation *)
            (2. *. m *. s, 1. -. s)
          else
            (* Inverse quadratic interpolation *)
            let q = ldexp fa (ea - ec) /. fc
            and r = ldexp fb (eb - ec) /. fc in
            (s *. (2. *. m *. q *. (q -. r) -. (b -. a) *. (r -. 1.)),
             (q -. 1.) *. (r -. 1.) *. (s -. 1.)) in
        let p, q = if p > 0. then p, -. q else -. p, q in
        (* If b+p/q falls in [b,c] and isn't too large, it is accepted *)
        if 2. *. p < 3. *. m *. q -. abs_float(tol_act *. q)
           && p < abs_float(0.5 *. e *. q)
        then p /. q, d
        else m, m
    in
    (* Adjust the step to be not less than tolerance *)
    let b' = b +. (if abs_float step > tol_act then step
                   else if m > 0. then tol_act else -. tol_act) in
    let fb', eb' = f b' in
    (* Adjust c for it to have a sign opposite to that of b *)
    if (fb' > 0.) = (fc > 0.) then (* => fb' * fb <= 0 *)
      let d = b' -. b in
      if (eb <= eb' && ldexp (abs_float fb) (eb - eb') < abs_float fb')
         || (eb > eb' && ldexp (abs_float fb') (eb' - eb) >= abs_float fb) then
        brent2_loop half_tol f b' fb' eb' b fb eb b' fb' eb' d d
      else
        brent2_loop half_tol f b fb eb b' fb' eb' b fb eb d d
    else (* => fb' * fc <= 0 *)
      if (ec <= eb' && ldexp (abs_float fc) (ec - eb') < abs_float fb')
         || (ec > eb' && ldexp (abs_float fb') (eb' - ec) >= abs_float fc) then
        brent2_loop half_tol f b' fb' eb' c fc ec b' fb' eb' step e'
      else
        brent2_loop half_tol f b fb eb b' fb' eb' c fc ec step e'
  )

let brent2 ?(tol=eps) f a b =
  if tol < 0. then invalid_arg "Root1D.brent2: tol < 0.";
  let fa, ea = f a and fb, eb = f b in
  if fa = 0. then a
  else if fb = 0. then b
  else if (fa < 0. && fb < 0.) || (fa > 0. && fb > 0.) then
    invalid_arg "Root1D.brent: f(a) and f(b) must have opposite signs"
  else
    let d = b -. a in
    if (ea <= eb && ldexp (abs_float fa) (ea - eb) < abs_float fb)
       || (ea > eb && ldexp (abs_float fb) (eb - ea) >= abs_float fa) then
      brent2_loop (0.5 *. tol) f  b fb eb  a fa ea  b fb eb  d d
    else
      brent2_loop (0.5 *. tol) f  a fa ea  b fb eb  a fa ea  d d


(* Local Variables: *)
(* compile-command: "make -k -C .." *)
(* End: *)
